contracts/deposit-contracts/contract/src/detail.rs

```diff
@@ -0,0 +1,110 @@
+        CallStackElement::StoredSession { account_hash, .. } => {
+            // Stored session code acts in account's context, so if stored session wants to interact
+            // with an CEP-18 token caller's address will be used.
+            Key::from(account_hash)
+        }
+        CallStackElement::StoredContract {
+            contract_package_hash,
+            ..
+        } => Key::from(contract_package_hash),
+    }
+}
+
+/// Traverse the callstack
+pub(crate) fn get_immediate_caller() -> Result<Key, DepositError> {
+    let call_stack = runtime::get_call_stack();
+    call_stack
+        .into_iter()
+        .rev()
+        .nth(1)
+        .map(call_stack_element_to_address)
+        .ok_or(DepositError::InvalidContext)
+}
+
+/// Gets [`URef`] under a name.
+pub(crate) fn get_uref(name: &str) -> URef {
+    let key = runtime::get_key(name)
+        .ok_or(ApiError::MissingKey)
+        .unwrap_or_revert();
+    key.try_into().unwrap_or_revert()
+}
+
+pub fn get_named_arg_size(name: &str) -> Option<usize> {
+    let mut arg_size: usize = 0;
+    let ret = unsafe {
+        ext_ffi::casper_get_named_arg_size(
+            name.as_bytes().as_ptr(),
+            name.len(),
+            &mut arg_size as *mut usize,
+        )
+    };
+    match api_error::result_from(ret) {
+        Ok(_) => Some(arg_size),
+        Err(ApiError::MissingArgument) => None,
+        Err(e) => runtime::revert(e),
+    }
+}
+
+/// Optional named args are used for access control
+pub fn get_optional_named_arg_with_user_errors<T: FromBytes>(
+    name: &str,
+    invalid: DepositError,
+) -> Option<T> {
+    match get_named_arg_with_user_errors::<T>(name, DepositError::Phantom, invalid) {
+        Ok(val) => Some(val),
+        Err(_) => None,
+    }
+}
+
+/// Read optional arg and propagate errors
+pub fn get_named_arg_with_user_errors<T: FromBytes>(
+    name: &str,
+    missing: DepositError,
+    invalid: DepositError,
+) -> Result<T, DepositError> {
+    let arg_size = get_named_arg_size(name).ok_or(missing)?;
+    let arg_bytes = if arg_size > 0 {
+        let res = {
+            let data_non_null_ptr = contract_api::alloc_bytes(arg_size);
+            let ret = unsafe {
+                ext_ffi::casper_get_named_arg(
+                    name.as_bytes().as_ptr(),
+                    name.len(),
+                    data_non_null_ptr.as_ptr(),
+                    arg_size,
+                )
+            };
+            let data =
+                unsafe { Vec::from_raw_parts(data_non_null_ptr.as_ptr(), arg_size, arg_size) };
+            api_error::result_from(ret).map(|_| data)
+        };
+        // Assumed to be safe as `get_named_arg_size` checks the argument already
+        res.unwrap_or_revert_with(DepositError::FailedToGetArgBytes)
+    } else {
+        // Avoids allocation with 0 bytes and a call to get_named_arg
+        Vec::new()
+    };
+
+    bytesrepr::deserialize(arg_bytes).map_err(|_| invalid)
```

Comment:
This will propagate a Casper APIError to the contract runtime. I did not write any of the code in detail.rs and it was used by Casper Labs in the cep standards.

Reply:
It MIGHT be propagated to ApiError, but it does not have to - depending of what called will do with error. So far I am not worried about losing error details, but rather about this error not being used in the code at all - is arguments handling done properly?

Reply:
Error handling has been improved in the latest commit.